clippy_lints/src/append_instead_of_extend.rs

```diff
@@ -0,0 +1,82 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for occurrences where one vector gets extended instead of append
+    ///
+    /// **Why is this bad?** Using `extend` instead of `append` is more precise and faster
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let mut a = vec![1, 2, 3];
+    /// let mut b = vec![4, 5, 6];
+    ///
+    /// // Bad
+    /// a.extend(b.drain(..));
+    ///
+    /// // Good
+    /// a.append(&mut b);
+    /// ```
+    pub APPEND_INSTEAD_OF_EXTEND,
+    perf,
+    "default lint description"
+}
+
+declare_lint_pass!(AppendInsteadOfExtend => [APPEND_INSTEAD_OF_EXTEND]);
+
+impl<'tcx> LateLintPass<'tcx> for AppendInsteadOfExtend {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if does_function_use_append(cx, expr) {
+            let mut applicability = Applicability::MachineApplicable;
+            if let ExprKind::MethodCall(_, _, exprs, _) = expr.kind {
+                if let ExprKind::MethodCall(_, _, src_args, _) = &exprs[1].kind {
```

Comment:
You use these two `if let` expressions only to get the arguments for the `snipped` function. You could also move the lint emission into the if chain where you already extracted and validated these arguments. So to line 78. That would make it more lenient and removed a bit of complexity from the implementation.

Another possibility would be to return an `Option<>` from the check function which returns the two expressions if the lint should be triggered.

Reply:
Used the first approach